cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFAOrderEvaluator.kt

```diff
@@ -0,0 +1,368 @@
+     * in [node]. The operator is retrieved from the map [nodesToOp] and is probably the name of the
+     * function called. If the call is neither a [MemberCallExpression] nor a [ConstructExpression],
+     * it probably calls a function which does not have a "base" (as it is the case for C). In that
+     * case, we try to look up the base in the map [thisPositionOfNode].
+     *
+     * The base is prefixed with [eogPath] in order to differentiate between different paths of
+     * execution in the control flow.
+     *
+     * If the base cannot be retrieved, or if the [node] is not considered by the analysis (i.e., no
+     * entry for [node] exists in the map [consideredBases]), the method returns `null`.
+     */
+    private fun getBaseAndOpOfNode(node: CallExpression, eogPath: String): Pair<String, String>? {
+        // The "base" node, on which the DFA is based on. Ideally, this is a variable declaration in
+        // the end.
+        var base =
+            when {
+                node is MemberCallExpression -> node.base
+                node is ConstructExpression -> node.astParent?.getSuitableDFGTarget()
+                node.hasThisPosition() ->
+                    node.getBaseOfCallExpressionUsingArgument(node.getThisPosition())
+                else -> {
+                    val dfgTarget = node.getSuitableDFGTarget()
+                    if (dfgTarget != null && dfgTarget is ConstructExpression) {
+                        dfgTarget.getSuitableDFGTarget()
+                    } else {
+                        dfgTarget
+                    }
+                }
+            }
+
+        if (base is DeclaredReferenceExpression && base.refersTo != null) {
+            base = base.refersTo
+        }
+
+        if (base != null && consideredBases.contains(base.id)) {
+            // We add the path as prefix to the base in order to differentiate between
+            // the different paths of execution which both can use the same base.
+            val prefixedBase = "$eogPath|${base.name}.${base.id}"
+            return Pair(prefixedBase, nodesToOp[node]!!)
+        }
+
+        if (base == null) {
+            log.warn("The base of a call expression must not be null.")
+        } else if (!consideredBases.contains(base.id)) {
+            log.info("Skipping call because the base $[base.id} is not considered.")
+        }
+
+        return null
+    }
+
+    private fun Node.addEogPath(path: String) {
+        nodeIDtoEOGPathSet.computeIfAbsent(this.id) { mutableSetOf() }.add(path)
+    }
+
+    private fun Node.getEogPaths(): Set<String>? {
+        return nodeIDtoEOGPathSet[this.id]
+    }
+
+    /**
+     * If it's not an object-oriented language, we need to retrieve the base in a different way.
+     * Usually, it's the argument at index 0 but here, it's more obscure.
+     */
+    private fun Node.getThisPosition(): Int {
+        return thisPositionOfNode[this]!!
```

Comment:
Avoid `!!`, instead an `Int?` should probably be returned which would also make `hasThisPosition` obsolete